src/App.js

```diff
@@ -23,49 +24,128 @@ const generateSquares = () => {
+  const [squareClicked, setSquareClicked] = useState([]);
+
+  const squaresCopy = [...squares];
+
+  const onClickCallback = (key) => {
+    if (!isGameRunning){
+      return;
+    }
+
+    for (let i=0; i<squareClicked.length; i++){
+      if (key === squareClicked[i]){
+        return;
+      }
+    }
+
+    for (let i = 0; i < 3; i++) {
+      for (let j = 0; j < 3; j++) {
+
+        if (squaresCopy[i][j].id == key) {
+          setSquareClicked(squareClicked.concat(key));
+          squaresCopy[i][j].value = currentPlayer;
+
+          if (currentPlayer === PLAYER_1) {
+            setCurrentPlayer(PLAYER_2);
+            setMessage(`Current player is: ${PLAYER_2}`);
+          } else if (currentPlayer === PLAYER_2) {
+            setCurrentPlayer(PLAYER_1);
+            setMessage(`Current player is: ${PLAYER_1}`);
+          }
+        }
+      }
+    }
+    setSquares(squaresCopy);
+
+    const winner = checkForWinner();
+    if (winner === 'x' || winner === 'o') {
+      setMessage(`Winner is ${winner}`);
+      setIsGameRunning(false);
+      return;
+    }
+    if (squareClicked.length === 8){
+      setMessage('The game ended with no winner. Click the reset button to play again');
+      setIsGameRunning(false);
+    }
+  };
 
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if 
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if 
-    //    all three squares have the same value.
+    let i = 0;
 
-  }
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
+  };
 
   const resetGame = () => {
-    // Complete in Wave 4
-  }
+    for (let i = 0; i < 3; i++) {
+      for (let j = 0; j < 3; j++) {
+        squaresCopy[i][j].value = ''; 
+      }
+    }
+    setSquares(squaresCopy);
```

Comment:
instead of doing this you could do `setSquares(generateSquares())` which will make the square values empty again